crates/circuit/src/packed_instruction.rs

```diff
@@ -39,132 +42,375 @@ enum PackedOperationType {
+        Self { instruction }
+    }
+}
+
+impl From<PointerBits> for BitField {
+    fn from(pointer: PointerBits) -> Self {
+        Self { pointer }
+    }
+}
+
+// Used by bitfield-struct to convert into and out of the bitfield.
+impl StandardGate {
+    const fn into_bits(self) -> u8 {
+        self as _
+    }
+    const fn from_bits(value: u8) -> Self {
+        if value as usize >= STANDARD_GATE_SIZE {
+            panic!("unexpected standard gate type!")
+        }
+        // It would've been nice to use bytemuck here, but it doesn't support const
+        // casting for enums.
+        unsafe { std::mem::transmute(value) }
+    }
+}
+
+// Used by bitfield-struct to convert into and out of the bitfield.
+impl StandardInstructionType {
+    const fn into_bits(self) -> u8 {
+        self as _
+    }
+
+    const fn from_bits(value: u8) -> Self {
+        match value {
+            0 => StandardInstructionType::Barrier,
+            1 => StandardInstructionType::Delay,
+            2 => StandardInstructionType::Measure,
+            3 => StandardInstructionType::Reset,
+            _ => panic!("unexpected standard instruction type!"),
+        }
+    }
+}
+
+impl PackedOperationType {
+    const fn from_bits(value: u8) -> Self {
+        match value {
+            0 => Self::StandardGate,
+            1 => Self::StandardInstruction,
+            2 => Self::PyGatePointer,
+            3 => Self::PyInstructionPointer,
+            4 => Self::PyOperationPointer,
+            _ => panic!("unexpected discriminant type!"),
+        }
+    }
+}
+
+/// The bitfield layout used for standard gates.
+#[bitfield(u64)]
+struct StandardGateBits {
+    #[bits(3, default = PackedOperationType::StandardGate, access = RO)]
+    discriminant: PackedOperationType,
+    #[bits(8)]
+    standard_gate: StandardGate,
+    #[bits(53)]
+    __: u64,
+}
+
+/// The bitfield layout used for standard instructions.
+#[bitfield(u64)]
+struct StandardInstructionBits {
+    #[bits(3, default = PackedOperationType::StandardInstruction, access = RO)]
+    discriminant: PackedOperationType,
+    #[bits(8)]
+    standard_instruction: StandardInstructionType,
+    #[bits(21)]
+    _pad1: u32,
+    #[bits(32)]
+    value: ImmediateValue,
+}
+
+/// An inline value present within [StandardInstructionBits] layouts used to store arbitrary
+/// data to be interpreted according to the encoded standard instruction.
+#[derive(Clone, Copy, Debug)]
+#[repr(transparent)]
+struct ImmediateValue(u32);
+
+impl ImmediateValue {
+    const fn into_bits(self) -> u32 {
+        self.0
+    }
+
+    const fn from_bits(value: u32) -> Self {
+        Self(value)
+    }
+
+    #[inline]
+    fn from_delay_unit(unit: DelayUnit) -> Self {
+        Self(unit as u32)
+    }
+
+    #[inline]
+    fn delay_unit(&self) -> DelayUnit {
+        ::bytemuck::checked::cast(self.0 as u8)
+    }
+
+    #[inline]
+    fn u32(&self) -> u32 {
+        self.0
     }
+}
```

Comment:
This `ImmediateValue` is largely indicative of the problem I meant above.  I feel like there's assumptions and invariants that need to be upheld between two places now, and that makes things a lot more difficult:

1. the maximum width of a `StandardInstruction` payload must be the same size as `ImmediateValue`.  This PR already violates that from a typing perspective (`Barrier(usize)`).
2. `ImmediateValue` has been marked as deriving `Clone` and `Copy`, except the actual type that's been punned to a `u32` might not logically allow the `Clone` and `Copy`.  Obviously `u32` would let us clone it unfairly too, but it's easier to reason about ownership semantics if the _actual_ owner is the one producing the new type.
3. (much more minor and future looking) why do we need to apply an arbitrary limit of 32 bits anyway?  What if there's a payload in the future that's 2x16 bits, and those don't need to be pushed up to the 32-bit boundary for aligned access anyway.  It feels a bit neater to me to give `StandardInstruction` the whole payload if it wants it, and say it's required to return a 64-bit value with the low three bits 0, then let it put things where it wants them.